chromium/background.js

```diff
@@ -674,5 +683,57 @@ chrome.runtime.onMessage.addListener(function(message, sender, sendResponse){
     updateState();
   } else if (message.type == "remove_rule") {
     removeRule(message.object);
+  } else if (message.type == "import_settings") {
+    import_settings(message.object).then(() => {
+      sendResponse(true);
+    });
   }
 });
+
+// Send a message to the embedded webextension bootstrap.js to get settings to import
+chrome.runtime.sendMessage("import-legacy-data", import_settings);
+
+/**
+ * Import extension settings (custom rulesets, ruleset toggles, globals) from an object
+ * @param settings the settings object
+ */
+async function import_settings(settings){
+  if(settings.changed){
+    // Load custom rulesets and add to storage
+    await new Promise(resolve => {
+      for(let ruleset of settings.custom_rulesets){
+        all_rules.addFromXml((new DOMParser()).parseFromString(ruleset, 'text/xml'));
+      }
+      storage.set({"legacy_custom_rulesets": settings.custom_rulesets}, resolve);
+    });
+
+    // Load all the ruleset toggles into memory and store
+    let rule_toggle_promises = [];
+    for(let ruleset_name in settings.rule_toggle){
+      localStorage[ruleset_name] = settings.rule_toggle[ruleset_name];
+
+      rule_toggle_promises.push((ruleset_name, ruleset_active => {
+        return new Promise(resolve => {
+          for(let host in all_rules.targets){
+            for(let ruleset of all_rules.targets[host]){
+              if(ruleset.name == ruleset_name){
+                ruleset.active = ruleset_active;
+                resolve(true);
+                return;
+              }
+            }
+          }
+          resolve(false);
+        })
+      })(ruleset_name, settings.rule_toggle[ruleset_name]));
+    }
+    await Promise.all(rule_toggle_promises);
+
+    all_rules.ruleCache = new Map();
+
+    // Set/store globals
+    storage.set({'httpNowhere': settings.prefs.http_nowhere_enabled});
```

Comment:
Is there a reason you provide a callback when calling set above, but not here? It seems like we should use the callback on both, to detect success or failure.